policy: Add datapath proxy port priority

jrajahalme opened the pull request:
Add support for listener priority by adding a datapath policymap `ProxyPortPriority` field.

Add a new test for HTTP policy on a port range and use the precedence field to make sure it is not overridden by a more-specific allow rule, e.g., on a single intersecting port.


```release-note
HTTP policies are now supported on port ranges.
```

jrajahalme committed:
policy: Add internal type 'mapStateMap'

Add type 'mapStateMap' for the internal map of mapStateEntries. This
helps reduce direct use of the type 'mapStateEntry', making it easier to
track.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Add test case for ingress overlapping redirect on a port range

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

aspsk commented on bpf/lib/common.h:
```diff
@@ -400,18 +400,23 @@ struct policy_entry {
 #define LPM_FULL_PREFIX_BITS (LPM_PROTO_PREFIX_BITS + 16)   /* protocol and dport specified */
 
+/* Deny precedence bit in 'precedence' */
+#define PRECEDENCE_DENY 1 << 7
```
I would add parenthesis, just in case: ```(1 << 7)```

jrajahalme replied:
Done!

aspsk reviewed (approved):
I've read my part, looks good. One minor comment

michi-covalent reviewed (approved):
ack on the cli change 💻

learnitall reviewed (approved):
LGTM for API changes!

squeed commented on pkg/policy/mapstate.go:
```diff
@@ -860,28 +762,53 @@ func (ms *mapState) insertWithChanges(key Key, entry mapStateEntry, features pol
-	if newEntry.IsDeny {
-		// Delete covered entries
+		// Delete covered entries of lower precedence, and
+		// same precedence deny entries if the keys are different
```
What is the case when NarrowerOrEqualKeys could return a different key?

jrajahalme replied:
It is typical for `NarrowerOrEqualKeys` to return different keys, that would be all the keys with narrower port (range). But it can also return a key that is the same, and in that case we need to merge the entries, but only if the precedence of the entries is the same. For example equal keys with same precedence originating from different rules/selectors, rule labels need to be merged. Could also be two proxy redirects on same precedence, here the tie breaker rule for proxy ports is used to consistently select one of them. Similar consideration for Auth requirement.

joestringer commented on pkg/policy/api/l4.go:
```diff
@@ -160,7 +160,7 @@ type Listener struct {
 	// +kubebuilder:validation:Maximum=100
 	// +kubebuilder:validation:Optional
-	Priority uint16 `json:"priority"`
+	Priority uint8 `json:"priority"`
```
This stood out to me as an API change that modifies the type of a field, gets into @cilium/sig-k8s territory, so maybe others can chime in.

My rough understanding of k8s partly informed by [this document on a different topic](https://docs.google.com/document/d/1KEdOXNQq-ZSg9YsQlrYZDE6NA3Bg1Czavg147k949xU/edit?tab=t.0#heading=h.917jnl85b89m) is that k8s under the hood does not use a specific schema version to store the underlying data for a custom resource, but rather it will store some value and depending on client schemas, attempt to marshal the value with individual clients differently based on the schema provided by the client... given that this value is limited to the range `1..100` it seems like this should actually work correctly for any client (v1.16 with uint16 or with this PR with uint8). From that perspective it seems weirdly maybe OK to change this value? 

I'm guessing that this change probably saves us from some awkward type conversions throughout the code so if we can safely make this change then sure. I would have liked to review such a change independently just to make sure it's not lost in the noise, but I guess this discussion thread should give sufficient input for whether we are OK with this individual change or not.

squeed replied:
that's right; for CRDs the k8s apiserver only cares about json. It doesn't have any knowledge of the Go types at all; the CRD itself only states:

```yaml
priority:
  maximum: 100
  minimum: 1
  type: integer
```

squeed replied:
Also, our Go API types are almost impossible to use correctly (can't find the bug right now), so I can guarantee nobody is using the Go types directly :-(.

joestringer commented on bpf/lib/policy.h:
```diff
@@ -111,20 +120,25 @@ __policy_can_access(const void *map, struct __ctx_buff *ctx, __u32 local_id,
 	 */
 
-	/* L3/L4 lookup. */
+	/* L3 lookup. */
```
nit: Maybe for a future PR we could do a pass over these comments to make it clear that this is referring to "L3+", ie an entry that likely includes the L3 property, maybe the L4 property, but possibly neither (wildcard/match all traffic). It took me a little while to process that this comment change is not acknowledging a functional change, just documenting the semantics differently. Given the range of possible results of the lookup I can see how either "L3" or "L3/L4" as the description of the properties may be misleading.

jrajahalme replied:
Makes sense, left for future work.

jrajahalme replied:
Ended up adding the comments when I needed to fix up something else!

joestringer commented on pkg/policy/mapstate.go:
```diff
@@ -347,69 +353,54 @@ func (ms *mapState) Lookup(key Key) (MapStateEntry, bool) {
+
+		// 2. Entries at the same precedence, deny wins
+		// Check for the L3 deny first to match the datapath behavior
+		if l3entry.IsDeny() {
```
Curious why there isn't a similar check for `L4entry.IsDeny()`?

jrajahalme replied:
Added!

joestringer commented on pkg/policy/types/entry.go:
```diff
@@ -0,0 +1,245 @@
+// Listener 'priority' is encoded in to the low 7 bits of 'precedence', inverted
+func NewMapStateEntry(deny bool, proxyPort uint16, priority uint8, authReq AuthRequirement) MapStateEntry {
+	if deny && (proxyPort != 0 || priority != 0 || authReq != 0) {
+		panic("Invalid parameters for NewMapStateEntry")
```
`panic` makes me incredibly nervous. I understand that attempting to configure deny and some non-deny feature at the same time doesn't make any sense, but putting panics into production code can end up having severe impacts and I'm not entirely confident that we are running tests with all permutations to ensure this is never triggered.

Deny takes precedence anyway, so if you're trying to avoid error checking we could alternatively do something like the following example (which should also be caught by CI given we detect & fail on warnings now).

```suggestion
		log.WithFields(...).Warning("BUG: Invalid parameters for NewMapStateEntry. Please report this bug to Cilium developers.")
		proxyPort = 0
		priority = 0
		authReq = 0
```

joestringer replied:
I guess current examples also use `hclog.Stacktrace()`, you could consider that too.

jrajahalme replied:
Replaced the panic with normalizing the parameters (without any logging). This is now mature enough for that :-)

joestringer commented on pkg/policy/types/entry.go:
```diff
@@ -0,0 +1,245 @@
+func (e *MapStateEntry) Merge(entry MapStateEntry) {
+	if entry.Precedence^e.Precedence >= PrecedenceDeny {
+		panic("Merge called for incompatible entries")
+	}
```
Please no random panics deep in random merge functions. If there's a chance that bad input comes to this function, maybe return an error up the stack to a point where we can make a reasonable choice about how to ignore some portion of policy state preparation, log the error, and gracefully continue under a "fail-safe" mode.

I've discussed `panic` in a bit more detail here: https://github.com/cilium/cilium/issues/33364

jrajahalme replied:
Removed the panic, was a leftover from maturing this change!

joestringer commented on pkg/policy/mapstate.go:
```diff
@@ -732,11 +631,18 @@ func (ms *mapState) addKeyWithChanges(key Key, entry mapStateEntry, changes Chan
 		datapathEqual = oldEntry.MapStateEntry == entry.MapStateEntry
-		oldEntry.merge(&entry)
+
+		// oldEntry.merge(&entry)
```
```suggestion
```

jrajahalme replied:
Removed!

joestringer commented on pkg/policy/mapstate.go:
```diff
@@ -766,8 +672,14 @@ func (ms *mapState) addKeyWithChanges(key Key, entry mapStateEntry, changes Chan
 	if entry, exists := ms.get(key); exists {
+		if entry.Precedence != precedence {
+			fmt.Printf("SKIPPING DELETE DUE TO MISMATCHInG PRECEDENCE: %v != %v\n",
+				entry.Precedence, precedence)
```
If the daemon needs to print messages, please use the logging library.

jrajahalme replied:
Was leftover from local testing, removed!

joestringer commented on pkg/policy/types/auth.go:
This PR would have benefited from splitting refactoring vs. functional changes - both for the content of this file and the `MapStateEntry`. Probably the main commit can be ~100/700 lines shorter and make it easier to review and bisect if anything goes wrong later.

If we can avoid moving the content of this file, it would reduce code churn and help with tracking git history too. Not a big deal either way, I'm guessing there's some motivation but it's just extra noise in an already huge PR.

jrajahalme replied:
Brought `types/auth.go` back to reduce churn :-)

joestringer commented on pkg/policy/mapstate.go:
```diff
@@ -858,28 +762,53 @@ func (ms *mapState) insertWithChanges(key Key, entry mapStateEntry, features pol
+			if v.Precedence > newEntry.Precedence ||
+				// New deny entry is also bailed due to different covering deny key
+				// of the same precedence, equal keys need to be merged
+				v.Precedence == newEntry.Precedence && k != newKey {
```
Does this need grouping to retain the current behavior?

```suggestion
				(v.Precedence == newEntry.Precedence && k != newKey) {
```

joestringer replied:
If yes, there are similar examples lower in the function.

jrajahalme replied:
I think these are now moot after I simplified this.

joestringer reviewed (dismissed):
The consolidation of feature bitflags internally between redirects and denies feels a bit cleaner here, seems like a nice cleanup to generalize the interaction between these features and the map state entries. I spotted a couple of places where we could maybe look at removing dedicated deny fields in favor of the new mechanism? But we can look into that following this PR, it's not critical to address now. This PR is big enough as it is.

There's a few issues like failing to use the logging library and introducing calls to `panic` that should be addressed before we merge this PR. Evidently we're overdue to introduce linters to pick these up in CI.

I ran out of steam towards the end of the PR, would feel a bit more confident with extra @cilium/sig-policy eyes (though I see @squeed was already taking a look). The relatively minimal changes to the unit tests are probably a good sign though that the current functionality is retained.

tklauser commented on pkg/policy/types/entry.go:
```diff
@@ -0,0 +1,149 @@
+		if priority > 0 {
+			if priority > MaxListenerPriority {
+				priority = MaxListenerPriority
+			}
```
Could be shortened to:

```suggestion
			priority = min(priority, MaxListenerPriority)
```

jrajahalme replied:
Done!

tklauser commented on pkg/policy/mapstate.go:
```diff
@@ -858,28 +758,43 @@ func (ms *mapState) insertWithChanges(key Key, entry mapStateEntry, features pol
 // deleted.
 //
 // Incremental changes performed are recorded in 'changes'.
-func (ms *mapState) denyPreferredInsertWithChanges(newKey Key, newEntry mapStateEntry, features policyFeatures, changes ChangeState) {
```
The godoc comment of [`(*mapState).revertChanges`](https://github.com/cilium/cilium/pull/36056/files#diff-171041eb8ba793ddbbd91b25b747b921c76878b41d78aabdc44a6633fea3409cR714) above and [`(*L4Filter).toMapState`](https://github.com/cilium/cilium/pull/36056/files#diff-d97453ecfcf9baa1157943e655c4e37306e9ac57daa1f7c7a67fcd37c2a82c4bL534) are still referring to `(*mapState).denyPreferredInsertWithChanges` and should probably be adjusted.

jrajahalme replied:
Done, thanks!

tklauser commented on pkg/policy/mapstate_test.go:
```diff
@@ -96,162 +92,146 @@ func (ms *mapState) validatePortProto(t *testing.T) {
 	return e
 }
 
 func TestMapState_denyPreferredInsertWithChanges(t *testing.T) {
```
This test should probably be renamed now that `(*mapState).denyPreferredInsertWithChanges` was renamed:

```suggestion
func TestMapState_insertWithChanges(t *testing.T) {
```

jrajahalme replied:
Done!

jrajahalme committed:
policy: Support listener priority on datapath

Add support for listener priority by adding a datapath policy map
ProxyPortPriority field.

This makes it possible to use HTTP policies with port ranges.

Move MapStateEntry definition into policy/types, so that it's components
are easier to import without causing merge conflicts.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

tklauser reviewed (approved):
Thanks
